proof/refine/ARM/Syscall_R.thy

```diff
@@ -552,33 +552,42 @@ lemma sts_mcpriority_tcb_at'[wp]:
 context begin interpretation Arch . (*FIXME: arch_split*)
 
 lemma sts_valid_inv'[wp]:
-  "\<lbrace>valid_invocation' i\<rbrace> setThreadState st t \<lbrace>\<lambda>rv. valid_invocation' i\<rbrace>"
-  apply (case_tac i, simp_all add: sts_valid_untyped_inv' sts_valid_arch_inv')
-         apply (wp | simp)+
-     defer
+  "setThreadState st t \<lbrace>valid_invocation' i\<rbrace>"
+  apply (case_tac i; simp)
+             apply (wpsimp wp: sts_valid_untyped_inv')
+            apply (wpsimp+)[4]
```

Comment:
It's not necessary, but it would be nice if we could work out a way to automate lines like this in this proof so that they stopped in the right spot. Failing that, would you be able to add comments indicating what the proof state should look like after each one?

This sort of already happens due to the structure of the proof with the case_tacs, but making it really clear would be nice.

Reply:
Also, I really like the new structure that you've given this proof! It might have ended up with more lines but I think it will be much more maintainable.

Reply:
Thanks! That's what I hope. I think that sometime more lines means that the exact point of failure is easier to find.